### ISSUE TITLE
Investigate why Xcode Clang-produced binaries fail to parse

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/llvm-pretty-bc-parser#235]
Multiple people have compiled this `swap.c` file (from [this SAW tutorial](https://saw.galois.com/intro/_downloads/b78063675d82bd43de0a313424ddfa57/examples.tar.gz)) using an Xcode version of Clang on macOS that crashes when loaded into `llvm-pretty-bc-parser`:

<details>

```hs
#include <stdlib.h>
#include <time.h>
#include <assert.h>

#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>

void swap(uint32_t *x, uint32_t *y) {
    uint32_t tmp = *x;
    *x = *y;
    *y = tmp;
}

bool swap_spec(uint32_t a, uint32_t b) {
    uint32_t x = a, y = b;
    swap(&x, &y);
    return x == b && y == a;
}

void xor_swap(uint32_t *x, uint32_t *y) {
    *x ^= *y;
    *y ^= *x;
    *x ^= *y;
}

bool xor_swap_spec(uint32_t a, uint32_t b) {
    uint32_t a1 = a, a2 = a;
    uint32_t b1 = b, b2 = b;
    xor_swap(&a1, &b1);
    swap(&a2, &b2);
    return a1 == a2 && b1 == b2;
}


bool general_swap_spec(void (*fun)(uint32_t *, uint32_t *), uint32_t a, uint32_t b) {
    uint32_t x = a, y = b;
    (*fun)(&x, &y);
    return x == b && y == a;
}

void swap_broken1(uint32_t *x, uint32_t *y) {
    uint32_t tmp = *x;
    *x = tmp;
    *y = *y;
}

bool swap_broken1_spec(uint32_t a, uint32_t b) {
    uint32_t x = a, y = b;
    swap_broken1(&x, &y);
    return x == b && y == a;
}


void swap_broken2(uint32_t *x, uint32_t *y) {
    uint32_t tmp = *x;
    if (*x != 4142351) {
        *x = *y;
        *y = tmp;
    }
}

bool swap_broken2_spec(uint32_t a, uint32_t b) {
    uint32_t x = a, y = b;
    swap_broken2(&x, &y);
    return x == b && y == a;
}


void swap_broken3(uint32_t *x, uint32_t *y) {
    uint32_t tmp = *x;
    if (*x && *x == (*y << 5)) {
        y = NULL;
    }
    *x = *y;
    *y = tmp;
}

bool swap_broken3_spec(uint32_t a, uint32_t b) {
    uint32_t x = a, y = b;
    swap_broken3(&x, &y);
    return x == b && y == a;
}



void test_swap_function(void (*fun)(uint32_t *, uint32_t *), char *descr, uint32_t x, uint32_t y) {
    printf("[%s] Testing with %u and %u... ", descr, x, y);
    if (general_swap_spec(fun, x, y)) {
        printf("OK\n");
    } else {
        printf("FAILED\n");
    }
}

void test_swap(char *descr, uint32_t x, uint32_t y) {
    test_swap_function(&swap, descr, x, y);
}

void chosen_value_test(void (*fun)(uint32_t *, uint32_t *)) {
    test_swap_function(fun, "Chosen", 1, 2);
    test_swap_function(fun, "Chosen", 2429, 98423);
    test_swap_function(fun, "Chosen", 8347853, 0);
    test_swap_function(fun, "Chosen", 5, 5);
    // ...
}

void random_value_test(void (*fun)(uint32_t *, uint32_t *)) {
    srand(time(NULL));

    for (int i = 0; i < 100; i ++) {
        test_swap_function(fun, "Random", rand(), rand());
    }
}

int main() {
    printf("Beginning chosen-value tests for swap\n");
    chosen_value_test(&swap);
    printf("\n");
    printf("Ending chosen-value tests for swap\n");

    printf("Beginning random tests for swap\n");
    random_value_test(&swap);
    printf("Ending random tests for swap\n");


}
```
</details>

```
[16:36:44.978] Loading file "/Users/mccleeary/dev/examples/intro/swap_cryptol.saw"
[16:36:44.986] Stack trace:
"llvm_load_module" (/Users/mccleeary/dev/examples/intro/swap_cryptol.saw:1:6-1:22)
Invalid record size: 15
Expected size between 12 and 14
Are you sure you're using a supported version of LLVM/Clang?
Check here: https://github.com/GaloisInc/llvm-pretty-bc-parser
```

As I do not run macOS, I am unable to easily reproduce this. I did, however, get an affected user to give me their `swap.bc` file (produced with `clang -g -O1 -c -emit-llvm swap.c -o swap.bc`), which I've attached here: [swap.bc.tar.gz](https://github.com/GaloisInc/llvm-pretty-bc-parser/files/12252468/swap.bc.tar.gz) I was also given the `swap.ll` file (produced with `clang -g -O1 -S -emit-llvm swap.c -o swap.ll`):  [swap.ll.tar.gz](https://github.com/GaloisInc/llvm-pretty-bc-parser/files/12252569/swap.ll.tar.gz)

I stared at the `swap.ll` file a bit, but nothing about it seems particularly suspicious. This part of the `llvm-pretty-bc-parser` error message, on the other hand, seems very suspicious:

```
Invalid record size: 15
Expected size between 12 and 14
```

This error is coming from this part of the code:

https://github.com/GaloisInc/llvm-pretty-bc-parser/blob/ac9fff49ef1670e58be9ce90075eaa02726f8662/src/Data/LLVM/BitCode/IR/Metadata.hs#L594-L595

So how are we ending up with a `METADATA_DERIVED_TYPE` record with 15 entries? I checked how LLVM itself parses `METADATA_DERIVED_TYPE` records [here](https://github.com/llvm/llvm-project/blob/f887cb10acc9fb21e389703062398adf1331232f/llvm/lib/Bitcode/Reader/MetadataLoader.cpp#L1540-L1542), and it _also_ enforces the convention that `METADATA_DERIVED_TYPE` has no more than 14 entries. Keep in mind that the code that I linked to is the most recent LLVM source (at the time of writing), so this is unlikely to be a new-LLVM-not-supported problem. (Indeed, we should support up to LLVM 16 already.)

As a result, I'm rather baffled as to how Xcode's Clang could produce something like this. Perhaps Xcode is using a forked version of LLVM that encodes metadata differently? In any case, I am stumped, so I am opening this issue as a reminder to look into this more deeply if I manage to get convenient access to Xcode for testing purposes.

**Comment by mccleeary-galois** [comment]
The following simple program reproduces the issue:

```c
typedef int myInt;

myInt main(){
    return 0;
}
```

The output of llvm-bcanalyzer from a .bc file of the above code compiled with Apple Clang V 14.0.3 produces the following for DERIVED_TYPE:
```
<DERIVED_TYPE op0=0 op1=22 op2=40 op3=37 op4=1 op5=0 op6=43 op7=0 op8=0 op9=0 op10=0 op11=0 op12=0 op13=0 op14=0/>
```
Which has more elements than the assert noted above by Ryan Scott.

**Comment by RyanGlScott** [comment]
For the time being, I am inclined to do something rather simple:

* Change [this lin(https://github.com/GaloisInc/llvm-pretty-bc-parser/blob/ac9fff49ef1670e58be9ce90075eaa02726f8662/src/Data/LLVM/BitCode/IR/Metadata.hs#L594-L595) to accept up to 15 records for `METADATA_DERIVED_TYPE`. This way, `llvm-pretty-bc-parser` won't choke on `.bc` files produced by recent Apple Clangs.
* If a `METADATA_DERIVED_TYPE` entry contains 15 records, simply ignore the last one, which (hopefully) corresponds to Apple's pointer authentication info.
 
This means that we wouldn't need any special Apple Clang–specific parsing logic, at least for the time being. If new Apple Clang versions move the pointer authentication info around to other places in the `METADATA_DERIVED_TYPE` entry (which seems very possible), we'd of course need to rethink this design.

**Comment by eddywestbrook** [comment]
That's great that you guys tracked down the source of the bug, but pretty frustrating that Apple would take a format that is supposed to be architecture independent and make their own architecture-dependent version of it. Frustrating but I guess not that surprising, coming from Apple. Your approach of doing a simple fix that seems to work for all the cases we've seen so far makes a lot of sense, @RyanGlScott, thanks for figuring that out!

**Comment by RyanGlScott** [comment]
I've submitted #247, which implements the idea in https://github.com/GaloisInc/llvm-pretty-bc-parser/issues/235#issuecomment-1677214307.

If `llvm-pretty-bc-parser` were more permissive when parsing LLVM metadata, we might have been able to avoid having tools like SAW crash when deployed on a version of Clang we hadn't explicitly tested yet. I've opened #248 to track this idea, although fixing that is not a prerequisite for fixing this issue.
